### PR TITLE
Some simple fixes

### PR DESCRIPTION
**Pull request by cyyever**

### CODE REVIEW

File: torch/csrc/autograd/python_cpp_function.cpp [hunk 1]
```diff
@@ -151,7 +151,9 @@ PyObject* THPCppFunction_metadata(THPCppFunction* self, void* _unused) {
   auto metadata =
       static_cast<PyAnomalyMetadata*>(self->cdata->metadata())->dict();
 
-  Py_INCREF(metadata);
+  if (metadata) {
+    Py_INCREF(metadata);
+  }
```
Comment (Skylion007) [inline torch/csrc/autograd/python_cpp_function.cpp:156]:
```suggestion
  Py_XINCREF(metadata);
```
is the idiomatic way of doing this in CPython.

**Review by Skylion007 — APPROVED** [review]
Other than these two minor style / perf fixes, this PR looks good.

File: torch/csrc/autograd/custom_function.cpp [hunk 2]
```diff
@@ -332,7 +332,7 @@ optional_variable_list _process_backward_mode_ad(
       var.mutable_grad().reset();
       impl::clear_hooks(var);
       if (auto grad_acc_fn = impl::try_get_grad_accumulator(var)) {
-        auto grad_acc = dynamic_cast<AccumulateGrad*>(grad_acc_fn.get());
+        auto grad_acc = std::dynamic_pointer_cast<AccumulateGrad>(grad_acc_fn);
```
Comment (Skylion007) [inline torch/csrc/autograd/custom_function.cpp:335]:
```suggestion
        auto grad_acc = std::dynamic_pointer_cast<AccumulateGrad>(std::move(grad_acc_fn));
```
Otherwise, this is slightly less efficient than the previous code since it requires a global lock to increment / decrement the smartpointer reference count.

Comment (Skylion007) [reply torch/csrc/autograd/custom_function.cpp:335]: Ah whoops, I misread the STL docs. This isn't in C++ until C++20. whoops.

Comment (Skylion007) [reply torch/csrc/autograd/custom_function.cpp:335]: This is still slightly more inefficient than what was there before, any reason why to do so?

File: torch/csrc/autograd/custom_function.cpp [hunk 3]
```diff
@@ -332,7 +332,7 @@ optional_variable_list _process_backward_mode_ad(
       var.mutable_grad().reset();
       impl::clear_hooks(var);
       if (auto grad_acc_fn = impl::try_get_grad_accumulator(var)) {
-        auto grad_acc = dynamic_cast<AccumulateGrad*>(grad_acc_fn.get());
+        auto grad_acc = std::dynamic_pointer_cast<AccumulateGrad>(std::move(grad_acc_fn));
```
Comment (Skylion007) [inline torch/csrc/autograd/custom_function.cpp:335]: @cyyever if you read the full comment chain, this doesn't actually work until C++20 so it will still be a bit of perf regression over the code that was previously there. Can you revert the changes to this file?

Comment (Skylion007) [reply torch/csrc/autograd/custom_function.cpp:335]: Or if you want to change it, it should be `auto *grad_acc =` ...

File: torch/csrc/autograd/python_cpp_function.cpp [hunk 4]
```diff
@@ -151,7 +151,7 @@ PyObject* THPCppFunction_metadata(THPCppFunction* self, void* _unused) {
   auto metadata =
```
Comment (Skylion007) [inline torch/csrc/autograd/python_cpp_function.cpp:151]: Also should be `auto *metadata`

**Comment by Skylion007** [comment]
Oh that linter failure is not from master. Regardless, should provide an opportunity to address the couple of lingering issues.

File: torch/csrc/autograd/custom_function.cpp (hunk 3 above)
Comment (cyyever) [reply torch/csrc/autograd/custom_function.cpp:335]: Can we discard the move? I want to use dynamic_pointer_cast to avoid dereference a possible nullptr

File: torch/csrc/autograd/python_cpp_function.cpp (hunk 4 above)
Comment (cyyever) [reply torch/csrc/autograd/python_cpp_function.cpp:151]: OK

**Comment by cyyever** [comment]
@Skylion007 see the last commit

**Review by Skylion007 — APPROVED** [review]
